plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/conversion/StmtConversionVisitor.kt

```diff
@@ -251,7 +251,12 @@ object StmtConversionVisitor : FirVisitor<ExpEmbedding, StmtConversionContext>()
     ): ExpEmbedding {
         val exp = data.convert(smartCastExpression.originalExpression)
         val newType = data.embedType(smartCastExpression.smartcastType.coneType)
-        return exp.withType(newType)
+        // If the smart-cast is from A? to A, then is not necessary to inhale invariants
+        return if (exp.type.getNonNullable() == newType) {
+            exp.withType(newType)
+        } else {
+            exp.withType(newType).withAccessInvariants()
+        }
```

Comment:
Would it be possible to express this logic a bit more flexibly?  I think that some kind of `invariants of A - invariants of B` operation for casting from `B` to `A` would be good.

Reply:
Yes, this is something we'll have to do with mutable predicates.
I think this can be done in two ways:
- find by ourselves `inv A - inv B` inhale them and then fold `A`. Here the difficult part is to find `inv A - inv B`
- inhaling and applying `B --* A && A == old[lhs](A)`. Here it is difficult to ensure that A is unchanged after applying the magic wand

Anyway, I think it's better to do this when we'll have mutable predicates. For now it's not a problem to inhale something twice

Reply:
Okay, could you make an issue and put a TODO for it here?